### ISSUE TITLE
Aggregation performance

### ISSUE THREAD

**Issue by mazswojejzony** [kofrasa/mingo#95]
Please see the following demo project: https://runkit.com/mazswojejzony/mingo-aggregation

In a nutshell it takes around 5 seconds to process aggregation logic on 10'000 items. 

Am I doing anything wrong here?

**Comment by kofrasa** [comment]
Hi @mazswojejzony, thanks for reporting this issue.

I haven't done extensive perf testing for `mingo` and have no data to support my comments so take it with a pinch of salt.

A 5 seconds processing time for 10,000 items is slow. My initial guess is that the time is spent within the `$group` pipeline operator, which requires realizing the full collection for the group expression operators `$sum`, `$max`, `$avg` etc. It is possible to optimize the partial collections (groups) processing by only computing the value of the objects once and only clone them in the operators if they will be modified.

I will experiment with that approach and see how that improves the benchmark. There are other optimizations that can be done as well but this looks like the cheapest to me now. I am open to more ideas.

Thanks for providing code samples with this report.

**Comment by kofrasa** [comment]
How does this compare to using another collection manipulation library?

**Comment by mazswojejzony** [comment]
Sorry, I have not checked other libraries. On the other hand I know the same aggregation pipeline runs way faster on MongoDB (I do not have exact numbers at hand though).

**Comment by mazswojejzony** [comment]
I've just reproduced the same scenario on my local MongoDB instance. The same aggregation pipeline runs within **0.1s** on my machine (Core i7 CPU, 16GB RAM, MongoDB running locally in a Docker container).

**Comment by kofrasa** [comment]
Hi @mazswojejzony, I was talking about comparing `mingo` against other collection filtering and aggregation javascript libraries like `lodash`, `Lazy.js` etc. `MongoDB` is implemented in C++ so any kind of comparison with `mingo` will have a disproportionate difference in the result.

**Comment by mazswojejzony** [comment]
Hi, I fully agree this is not apples to apples comparison. At the same time I think it gives rough idea on what is possible.

**Comment by darknos** [comment]
I compared my pipeline with 1.x version of mingo and got result almost 3x times faster. it’s a bit confusing

**Comment by kofrasa** [comment]
I have tracked down the change that introduced the slow performance to [86b13fd](https://github.com/kofrasa/mingo/commit/86b13fdb2f3173d0f9aa5a54dbead065157220ba#diff-3deb3f32958bb937ae05c6f3e4abbdf5) which was introduced between releases `1.3.3` and `2.0.0`.  

The implementations for `inArray` and `notInArray` were changed to rely on `isEqual` which performs a deep equality check that is unnecessary in most places where is it used. I have reverted to using `Array.includes` and use the deep equality check explicitly for when it is required.

The performance is looking sensible again. Between `800ms` and `900ms` on my 2.6GHz i5.
